datumaro/components/converter.py

```diff
@@ -34,8 +36,22 @@ def convert(cls, extractor, save_dir, **options):
         return converter.apply()
 
     @classmethod
+    @error_rollback
     def patch(cls, dataset, patch, save_dir, **options):
-        return cls.convert(dataset, save_dir, **options)
+        # Is not any better in performance than just writing a dataset,
+        # but in case of patching (i.e. writing to the previous location),
+        # it allows to avoid many problems with removing and replacing
+        # existing files. Surely, this approach also has problems with
+        # removal of the given directory.
+        tmpdir = save_dir + '.tmp' + uuid.uuid4().hex
+        os.makedirs(tmpdir, exist_ok=False)
+        on_error_do(shutil.rmtree, tmpdir, ignore_errors=True)
+
+        retval = cls.convert(dataset, tmpdir, **options)
+        shutil.rmtree(save_dir, ignore_errors=True)
```

Comment:
Ignoring errors here seems like a bad idea. If any errors occur, the following `os.replace` call will fail anyway, so you might as well fail early.

Reply:
Changed